hugr-core/src/import.rs

```diff
@@ -1138,62 +1144,86 @@ impl<'a> Context<'a> {
         term_id: model::TermId,
     ) -> Result<FuncTypeBase<RV>, ImportError> {
         let (inputs, outputs, extensions) = self.get_func_type(term_id)?;
-        let inputs = self.import_type_row::<RV>(inputs)?;
-        let outputs = self.import_type_row::<RV>(outputs)?;
+        let inputs = self.import_type_row(inputs)?;
+        let outputs = self.import_type_row(outputs)?;
         let extensions = self.import_extension_set(extensions)?;
         Ok(FuncTypeBase::new(inputs, outputs).with_extension_delta(extensions))
     }
 
     fn import_closed_list(
         &mut self,
-        mut term_id: model::TermId,
+        term_id: model::TermId,
     ) -> Result<Vec<model::TermId>, ImportError> {
-        // PERFORMANCE: We currently allocate a Vec here to collect list items
-        // into, in order to handle the case where the tail of the list is another
-        // list. We should avoid this.
-        let mut list_items = Vec::new();
+        let mut types = Vec::new();
+        let mut stack = vec![term_id];
 
-        loop {
+        while let Some(term_id) = stack.pop() {
             match self.get_term(term_id)? {
-                model::Term::Var(_) => return Err(error_unsupported!("open lists")),
-                model::Term::List { items, tail } => {
-                    list_items.extend(items.iter());
-
-                    match tail {
-                        Some(tail) => term_id = *tail,
-                        None => break,
+                model::Term::List { items } => {
+                    types.reserve(items.len());
```

Comment:
Is this worth it? It's a conservative overestimate and we'll be doing it repeatedly